jgrapht-core/src/main/java/org/jgrapht/alg/cycle/QueueBFSFundamentalCycleBasis.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * (C) Copyright 2016-2017, by Dimitrios Michail and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.cycle;
+
+import java.util.ArrayDeque;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.jgrapht.Graph;
+import org.jgrapht.Graphs;
+
+/**
+ * Generate a set of fundamental cycles by building a spanning tree (forest) using a straightforward
+ * implementation of BFS using a FIFO queue. The implementation first constructs the spanning forest
+ * and then builds the fundamental-cycles set. It supports graphs with self-loops and/or graphs with
+ * multiple edges.
+ * 
+ * <p>
+ * For information on algorithms computing fundamental cycle bases see the following paper: Narsingh
+ * Deo, G. Prabhu, and M. S. Krishnamoorthy. Algorithms for Generating Fundamental Cycles in a
+ * Graph. ACM Trans. Math. Softw. 8, 1, 26-42, 1982.
+ * </p>
+ * 
+ * <p>
+ * The total length of the fundamental-cycle set can be as large as O(n^3) where n is the number of
+ * vertices of the graph.
+ * </p>
+ * 
+ * @param <V> the vertex type
+ * @param <E> the edge type
+ *
+ * @author Dimitrios Michail
+ * @since October 2016
+ */
+public class QueueBFSFundamentalCycleBasis<V, E>
```

Comment:
The version with the stack is more similar, but still class `PatonCycleBase` produces a different cycle basis. 

It does not work with multigraphs, but it produces a "weakly" fundamental cycle basis which in some cases might be shorter that the ones returned by the other algorithms. When this happens, it runs slightly faster (same complexity, less constants).